client/post-editor/term-token-field/index.jsx

```diff
@@ -0,0 +1,103 @@
+/**
+ * External dependencies
+ */
+import React from 'react';
+import cloneDeep from 'lodash/cloneDeep';
+import { connect } from 'react-redux';
+import _debug from 'debug';
+
+/**
+ * Internal dependencies
+ */
+import { getSelectedSiteId } from 'state/ui/selectors';
+import { getTerms } from 'state/terms/selectors';
+import TokenField from 'components/token-field';
+import { decodeEntities } from 'lib/formatting';
+import TermsConstants from 'lib/terms/constants';
+import PostActions from 'lib/posts/actions';
+import { recordStat, recordEvent } from 'lib/posts/stats';
+import QueryTerms from 'components/data/query-terms';
+
+const debug = _debug( 'calypso:post-editor:editor-terms' );
+
+class TermTokenField extends React.Component {
+	componentWillMount() {
+		this.boundOnTermsChange = this.onTermsChange.bind( this );
+	}
+
+	onTermsChange( selectedTerms ) {
+		debug( 'onTermsChange', selectedTerms, this );
+
+		let termStat, termEventLabel;
+		if ( selectedTerms.length > this.getPostTerms().length ) {
+			termStat = 'term_added';
+			termEventLabel = 'Added Term';
+		} else {
+			termStat = 'term_removed';
+			termEventLabel = 'Removed Term';
+		}
+		recordStat( termStat );
+		recordEvent( 'Changed Terms', termEventLabel );
+
+		const { postTerms, taxonomyName } = this.props;
+		const terms = cloneDeep( postTerms ) || {};
+		terms[ taxonomyName ] = selectedTerms;
+		// TODO: REDUX - remove flux actions when whole post-editor is reduxified
+		PostActions.edit( { terms } );
+	}
+
+	getPostTerms() {
+		const { postTerms, taxonomyName } = this.props;
+
+		if ( ! postTerms || ! postTerms[ taxonomyName ] ) {
+			return [];
+		}
+
+		if ( Array.isArray( postTerms[ taxonomyName ] ) ) {
+			return postTerms[ taxonomyName ];
+		}
+
+		return Object.keys( postTerms[ taxonomyName ] );
+	}
+
+	render() {
+		const termNames = ( this.props.terms || [] ).map( term => term.name );
```

Comment:
Maybe I'm a bit Lodash obsessed, but the Lodash `map` implementation could simplify this a bit, removing the need for fallback array and callback function:

```js
const termNames = map( this.props.terms, 'name' );
```